File: Services/Teacher/Teacher.Application/Handler/GetCoursesQueryHandler.vb
Imports System.Threading
Imports MediatR
Imports Teacher.Core

Public Class GetCoursesQueryHandler
    Implements IRequestHandler(Of GetCoursesQuery, IEnumerable(Of CourseDto))

    Private ReadOnly _courseRepository As ICourseRepository
    Public Sub New(courseRepository As ICourseRepository)
        _courseRepository = courseRepository
    End Sub

    Public Async Function Handle(request As GetCoursesQuery, cancellationToken As CancellationToken) As Task(Of IEnumerable(Of CourseDto)) Implements IRequestHandler(Of GetCoursesQuery,IEnumerable(Of CourseDto)).Handle
        Dim courses = Await _courseRepository.GetAll()
        Return courses.Select(Function(course) New CourseDto With{
                                 .Id = course.Id,
                                 .Name = course.Name,
                                 .Description = course.Description
                                 })
    End Function
End Class

File: Services/Student/Student.Application/Handlers/GetCoursesQueryHandler.cs
using AutoMapper;
using MediatR;
using Microsoft.Extensions.Logging;
using Student.Application.Queries;
using Student.Application.Dtos;
using Student.Core.Repositories;

namespace Student.Application.Handlers;

public class GetCoursesQueryHandler : IRequestHandler<GetCoursesQuery, IEnumerable<CourseDto>>
{
    private readonly IMapper _mapper;
    private readonly ILogger<GetCourseQueryHandler> _logger;
    private readonly ICourseRepository _courseRepository;
    public GetCoursesQueryHandler(IMapper mapper, ILogger<GetCourseQueryHandler> logger, ICourseRepository courseRepository)
    {
        _mapper = mapper;
        _logger = logger;
        _courseRepository = courseRepository;
    }

    public async Task<IEnumerable<CourseDto>> Handle(GetCoursesQuery request, CancellationToken cancellationToken)
    {
        var courses = await _courseRepository.GetAll();
        if (!courses.Any()) return null!;
        return courses.Select(_mapper.Map<CourseDto>);

    }
}
